changefeedccl: check compatibility between sinks and changefeed options

spiffyy99 opened the pull request:
Add checks for option compatibility with sinks (separated into global
options and sink-specific options).

Also converted the avro encoder tests to use Kafka sink, removes the
need to create separate schema registries and convert Avro payloads to
JSON within tests, since Kafka testfeed does this automatically.

Resolves cockroachdb#68647

Release note (enterprise change): Changefeeds will now error if an
option is used with an incompatible sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69173)
<!-- Reviewable:end -->

spiffyyeng commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -162,7 +162,8 @@ func testingSetMaxLogIgnoredImportStatements(maxLogSize int) (cleanup func()) {
 
-func makeStringSet(opts ...string) map[string]struct{} {
+// MakeStringSet returns a set of strings
+func MakeStringSet(opts ...string) map[string]struct{} {
```
not sure if there's a better place to relocate this to

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -8,7 +8,10 @@
 
-import "github.com/cockroachdb/cockroach/pkg/sql"
+import (
+	"github.com/cockroachdb/cockroach/pkg/ccl/importccl"
```
Let's not add this dependency.  Code reuse is good an all, but we're probalby importing a small utility function -- we can reimplement it ourselves.

spiffyyeng replied:
duplicated in `options.go`

miretskiy commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -162,7 +162,8 @@ func testingSetMaxLogIgnoredImportStatements(maxLogSize int) (cleanup func()) {
 }
 
-func makeStringSet(opts ...string) map[string]struct{} {
+// MakeStringSet returns a set of strings
```
It's a tiny function -- we should just implement it ourselves.

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -160,3 +163,19 @@ var ChangefeedOptionExpectValues = map[string]sql.KVStringOptValidate{
 }
+
+// ChangefeedOptionsAllSinks is options common to all sinks
+var ChangefeedOptionsAllSinks = importccl.MakeStringSet(OptCursor, OptEnvelope,
```
I think we know we're dealing w/ changefeed options.
I would just shorten this var, as well as ones below:
CommonOptions, KafkaValidOptions, ChangefeedValidOptions.

miretskiy commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -312,6 +312,12 @@ func makeCloudStorageSink(
 	user security.SQLUsername,
 ) (Sink, error) {
+	// cloud storage sink has no sink-specific options
+	err := validateSinkOptions(opts, nil)
```
Let's add a variable (= nil) for all sink types so that everything is consistent.

spiffyyeng replied:
added cloudstorage (which is non-nil now) and SQL.

miretskiy commented on pkg/ccl/changefeedccl/sink.go:
```diff
@@ -124,6 +124,21 @@ func getSink(
 	return sink, nil
 }
 
+func validateSinkOptions(opts map[string]string, sinkSpecificOpts map[string]struct{}) error {
```
How would you feel about calling this function from the `getSink` call above so that all validation happens in 1 place?
You could add another helper in `getSink` which takes sink specific options, and a
call back responsible for actually creating the sink:

```
validateOptionsAndMakeSink := func(sinkOpts map[string]struct, makeSink func() (Sink, error)) (Sink, error) {
    if validateSinkOptions....; err != nil {
          
    }
    return makeSink()
}
```

Then, change newSink switch statement to delegate to that helper:
```
newSink .... {
    switch {
        ....
        case u.Scheme == changefeedbase.SinkSchemeKafka:
			return validateAndMakeSink(changefeedbase.KafkaOptions, func() (Sink, error) {
                return makeKafkaSink(ctx, sinkURL{URL: u}, feedCfg.Targets, feedCfg.Opts) 
            })
    }
}

spiffyyeng commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -224,34 +224,6 @@ func avroToJSON(t testing.TB, reg *cdctest.SchemaRegistry, avroBytes []byte) []b
 	return json
 }
 
-func assertPayloadsAvro(
```
removed this since we're now only using Avro tests with Kafka testfeed which automatically converts the payloads to JSON. If other sinks adopt Avro in the future we can add this back or perform the conversion within the corresponding testfeed (similar to Kafka).

spiffyyeng commented on pkg/ccl/changefeedccl/encoder_test.go:
```diff
@@ -831,30 +786,25 @@ func TestAvroMigrateToUnsupportedColumn(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
-		reg := cdctest.StartTestSchemaRegistry()
```
also don't need a separate schema registry since kafkafeed creates one automatically which will override this one. if we need to check the sink's registry we can directly access it as a member of kafkafeed.

spiffyy99 committed:
changefeedccl: check compatibility between sinks and changefeed options

Add checks for option compatibility with sinks (separated into global
options and sink-specific options).

Also converted the avro encoder tests to use Kafka sink, removes the
need to create separate schema registries and convert Avro payloads to
JSON within tests, since Kafka testfeed does this automatically.

Resolves #68647

Release justification: Simple error check that doesn't change any
changefeed behavior.

Release note (enterprise change): Changefeeds will now error if an
option is used with an incompatible sink.

Release justification:

spiffyyeng commented:
bors r+

spiffyyeng commented:
bors retry
